subprojects/composite-builds/src/main/java/org/gradle/composite/internal/IncludedBuildDependencySubstitutionsBuilder.java

```diff
@@ -39,6 +42,7 @@
     private final ImmutableAttributesFactory attributesFactory;
     private final NotationParser<Object, ComponentSelector> moduleSelectorNotationParser;
     private final NotationParser<Object, Capability> capabilitiesParser;
+    private final Set<IncludedBuildState> processed = new HashSet<>();
```

Comment:
Just a note: `IncludedBuildState` implementations do not implement equals/hashCode and we'll be relying on reference equality here. Not sure if this can lead to trouble in this case.
Alternatively, we could store `BuildIdentifiers` in this set.

Reply:
Thanks. Relying on object reference is fine here.